File: opensource/kernel/src/test/java/com/sos/scheduler/engine/kernel/plugin/PluginAdapterTest.java
package com.sos.scheduler.engine.kernel.plugin;

import com.google.inject.Injector;
import com.sos.scheduler.engine.kernel.log.PrefixLog;
import com.sos.scheduler.engine.kernel.scheduler.PrefixLogMock;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static com.google.inject.Guice.createInjector;
import static com.sos.scheduler.engine.kernel.util.XmlUtils.loadXml;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;

public final class PluginAdapterTest {
    private static final Logger logger = LoggerFactory.getLogger(PluginAdapterTest.class);
    private static final PluginConfiguration pluginConfiguration = new PluginConfiguration(PluginMock.class.getName(), ActivationMode.activateOnStart, null);
    private static final PluginConfiguration errorPluginConfiguration = new PluginConfiguration(ErrorPluginMock.class.getName(), ActivationMode.activateOnStart, null);

    private final PrefixLog log = PrefixLogMock.newLog(logger);
    private final PluginAdapter pluginAdapter = new PluginAdapter(pluginConfiguration, log);
    private final PluginAdapter errorPluginAdapter = new PluginAdapter(errorPluginConfiguration, log);
    private final Injector injector = createInjector();

    @Test public void testActivate() {
        pluginAdapter.activate(injector);
    }

    @Test(expected=RuntimeException.class) public void testActivate_error() {
        errorPluginAdapter.activate(injector);
    }

    @Test public void testGetXmState() {
        loadXml(pluginAdapter.getXmlState());
        errorPluginAdapter.getXmlState();
    }

    @Test public void testTryActivate() {
        pluginAdapter.tryActivate(injector);
        errorPluginAdapter.tryActivate(injector);
    }

    @Test public void testTryClose() {
        pluginAdapter.tryClose();
        errorPluginAdapter.tryClose();
    }

    @Test public void testGetXmlState() {
        loadXml(pluginAdapter.getXmlState());
        loadXml(errorPluginAdapter.getXmlState());
    }

    @Test public void testToString() {
        assertThat(pluginAdapter.toString(), startsWith("Plugin "));
    }
}

File: opensource/kernel/src/test/scala/com/sos/scheduler/engine/kernel/plugin/PluginAdapterTest.scala
package com.sos.scheduler.engine.kernel.plugin

import PluginAdapterTest._
import com.google.inject.Guice._
import com.google.inject.Injector
import com.sos.scheduler.engine.common.scalautil.Logger
import com.sos.scheduler.engine.common.xml.XmlUtils.loadXml
import com.sos.scheduler.engine.kernel.scheduler.PrefixLogMock
import java.lang.RuntimeException
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.startsWith
import org.junit.runner.RunWith
import org.scalatest.FunSuite
import org.scalatest.Matchers._
import org.scalatest.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
final class PluginAdapterTest extends FunSuite {
  private val log = PrefixLogMock.newLog(logger.delegate)
  private val pluginAdapter = new PluginAdapter(pluginConfiguration)
  private val errorPluginInstanceAdapter = new PluginAdapter(errorPluginConfiguration)
  private val injector: Injector = createInjector()

  test("initialize") {
    pluginAdapter.initialize(injector, log)
  }

  test("activate") {
    pluginAdapter should not be 'active
    pluginAdapter.activate()
    pluginAdapter should be ('active)
  }

  test("activate error") {
    errorPluginInstanceAdapter.initialize(injector, log)
    intercept[RuntimeException] { errorPluginInstanceAdapter.activate() }
  }

  test("tryActivate") {
    pluginAdapter should be ('active)
    pluginAdapter.tryActivate()
    pluginAdapter should be ('active)
    errorPluginInstanceAdapter.tryActivate()
    errorPluginInstanceAdapter should not be 'active
  }

  test("tryClose") {
    pluginAdapter should be ('active)
    pluginAdapter.tryClose()
    pluginAdapter should not be 'active
    errorPluginInstanceAdapter should not be 'active
    errorPluginInstanceAdapter.tryClose()
    errorPluginInstanceAdapter should not be 'active
  }

  test("xmlState") {
    loadXml(pluginAdapter.xmlState)
    loadXml(errorPluginInstanceAdapter.xmlState)
  }

  test("toString") {
    assertThat(pluginAdapter.toString, startsWith("Plugin "))
  }
}

private object PluginAdapterTest {
  val logger = Logger(getClass)
  val pluginConfiguration = new PluginConfiguration(classOf[MockPlugin].getName, ActivationMode.activateOnStart, None)
  val errorPluginConfiguration = new PluginConfiguration(classOf[ErrorMockPlugin].getName, ActivationMode.activateOnStart, None)
}

